app/utils/path_utils.py

```
@@ -75,6 +76,99 @@ def is_within_configured_volumes(path: str, *, treat_archives: bool = True) -> b
+    prev = None
+    while tail != prev:
+        prev = tail
+        tail = tail.rstrip(seps)
+        # Drop a standalone trailing "." component (".", ".../.") — but never the
+        # "." inside a ".." component, which stays meaningful.
+        if tail.endswith(".") and (len(tail) == 1 or tail[-2] in seps):
+            tail = tail[:-1]
+    return (drive + tail) or path
+
+
+def is_safe_directory_tree(path: str) -> bool:
```

Ruff (0.15.20)

[warning] Too many return statements (9 > 6)

(PLR0911)